### ISSUE TITLE
Google Ads query function does return INTERNAL ERROR ENCOUNTERED

### ISSUE THREAD

**Issue by BeautifulCodeStar** [googleads/google-ads-php#80]
Greetings Ads Team!

I have used Google Ads PHP library v0 and it was worked well.
Since version 0 is deprecated a few days ago, I have upgraded version to v1.

After that, all Google ads query function does not work and always return following exception and can't find specific issue related to this one.

`{↵    "message": "Internal error encountered.",↵    "code": 13,↵    "status": "INTERNAL",↵    "details": []↵}`

Is this Ads API internal issue or Something wrong on my side?

Thank you

**Comment by fiboknacky** [comment]
Hello,

Is that all you got as an error message? Are there any log messages too?

For v1, the only known error I can think of would be [`MISSING_TOS`](https://developers.google.com/google-ads/api/reference/rpc/google.ads.googleads.v1.errors), it happens when you've not sign the terms of service yet. They can be found here: ads.google.com/aw/apicenter

**Comment by BeautifulCodeStar** [comment]
Hello
Thanks for your support!

Yes, That's all what I got as exception message. 
No log found in API log file. That means it's not even called Google Ads API properly.

Regarding `MISSING_TOS`, YES. I followed that terms as well. I will check it again.

This issue is happening for over 1 week, I'm struggling to solve this issue.

Like another API call such as Ads account Authorization worked well. But not for all query functions.

Is there any other possible issues in this case?

**Comment by fiboknacky** [comment]
Could you share request ID too please?
Besides, what's the code you ran?

**Comment by BeautifulCodeStar** [comment]
Yes, its in PHP. Below issue details.

````````````
Method Name: /google.ads.googleads.v1.services.GoogleAdsService/Search
Host: googleads.googleapis.com
Headers: {
    "x-goog-api-client": "gl-php\/7.2.17-0ubuntu0.18.04.1 gapic\/ gax\/0.38.0 grpc\/1.4.0",
    "x-goog-request-params": "customer_id=XXXXXXXXXX",
    "developer-token": "REDACTED",
    "login-customer-id": "XXX-XXX-XXX"
}
Request: {"customerId":"7509260745","query":"SELECT campaign.id, campaign.name, metrics.ctr, metrics.clicks, metrics.impressions, metrics.average_cpc, metrics.cost_micros, metrics.average_cost, metrics.conversions\n                FROM campaign\n                LIMIT 50\n                ","pageSize":50}

Response
-------
Headers: {
    "date": "Thu, 16 May 2019 08:23:46 GMT",
    "alt-svc": "quic=\":443\"; ma=2592000; v=\"46,44,43,39\""
}

Fault
-------
Status code: 13
Details: Internal error encountered.
Failure: {}  
`````````````

**Comment by BeautifulCodeStar** [comment]
[2019-05-16 10:23:46] google-ads.INFO: Request made: Host: "googleads.googleapis.com", Method: "/google.ads.googleads.v1.services.GoogleAdsService/Search", ClientCustomerId: XXX, RequestId: "", IsFault: 1, FaultMessage: "None"  
[2019-05-16 10:23:46] google-ads.DEBUG: Request
-------

RequestId is displayed as empty string "".

**Comment by fiboknacky** [comment]
Looks like what you're trying to access (customer ID 750XXX) doesn't have any parent account, but you specified a customer ID in the login-customer-id header, correct?
In that case, it'd result in the permission denied error.

**Comment by BeautifulCodeStar** [comment]
This is some of code base for API call.
``````
$googleAdsServiceClient = $this->googleAdsClient->getGoogleAdsServiceClient();
        $query = "SELECT campaign.id, campaign.name, campaign.campaign_budget, campaign_budget.amount_micros, campaign.status, metrics.clicks, metrics.impressions, metrics.cost_micros, metrics.conversions FROM campaign ";
        if ($startDate && $endDate) {
            $query .= "WHERE segments.date BETWEEN '$startDate' AND '$endDate' ORDER BY campaign.id LIMIT 50";
        } else if ($filter) {
            $query .= "WHERE segments.date DURING " .$filter. " ORDER BY campaign.id LIMIT 50";
        } else {
            $query .= "WHERE campaign.status = ENABLED ORDER BY campaign.id LIMIT 50";
        }
        try {
            $options = ['pageSize' => self::PAGE_SIZE];
            if (!empty($pageToken)) {
                $options['pageToken'] = $pageToken;
            }
            $response = $googleAdsServiceClient->search($customerId, $query, $options);
        } catch(Exception $e) {
            $this->errorResponse($e);
        }
````

**Comment by BeautifulCodeStar** [comment]
FYI, I just deleted real customer ID for personal credential.

**Comment by BeautifulCodeStar** [comment]
I have confirmed that new `TOS` accepted against MCC and developer token again.

**Comment by fiboknacky** [comment]
Could you check what I replied in [this comment](https://github.com/googleads/google-ads-php/issues/80#issuecomment-492997519)?
Thank you.

**Comment by BeautifulCodeStar** [comment]
Yes.
In the header, there were customerID and MCC account ID.

And it's authorized via OAuth2. Hence it's not result in permission error and return INTERNAL SERVER ERROR always.

**Comment by fiboknacky** [comment]
What I meant is that your account isn't linked to any parent accounts, so you shouldn't use login-customer-id.
If you use login-customer-id, make sure it's the same as your account and your refresh token is also generated from the Google account that has access to that Ads account.

Could you confirm please?

**Comment by BeautifulCodeStar** [comment]
Yes, I confirm.

That's working with such way.

As i mentioned above, it's worked well so far.

Either my ads account or other ads accounts are not working properly after upgrading version1.

**Comment by BeautifulCodeStar** [comment]
I tried to pull fresh v1 repositories from Github and run examples such as GetCampigns.php.

That is also causing same Issue `INTERNAL SERVER ERROR` with status code 13.

**Comment by fiboknacky** [comment]
One thing you could try: delete your vendor/ folder and re-install all the dependencies again. 
There are probably some changes that may not be updated yet.

And I actually cannot find your log at all.
The closest one that I found is when you tried to access the customer ID 750XXX, with login-customer-id of another unrelated customer, which gave authentication error.

Could you try re-install the dependencies and let us know the result?

**Comment by BeautifulCodeStar** [comment]
Yes, Already I've tried that(delete vendor and reinstall) several times. And got same issue.

**Comment by BeautifulCodeStar** [comment]
I've listed completed log again.

```
[2019-05-13 03:45:33] google-ads.INFO: Request made: Host: "googleads.googleapis.com", Method: "/google.ads.googleads.v1.services.GoogleAdsService/Search", ClientCustomerId: 7509260745, RequestId: "", IsFault: 1, FaultMessage: "None"  
[2019-05-13 03:45:33] google-ads.DEBUG: Request
-------
Method Name: /google.ads.googleads.v1.services.GoogleAdsService/Search
Host: googleads.googleapis.com
Headers: {
    "x-goog-api-client": "gl-php\/7.2.17-0ubuntu0.18.04.1 gapic\/ gax\/0.38.0 grpc\/1.4.0",
    "x-goog-request-params": "customer_id=7509260745",
    "developer-token": "REDACTED",
    "login-customer-id": "992-748-0321"
}
Request: {"customerId":"7509260745","query":"SELECT campaign.id, campaign.name, metrics.ctr, metrics.clicks, metrics.impressions, metrics.average_cpc, metrics.cost_micros, metrics.average_cost, metrics.conversions\n                FROM campaign\n                LIMIT 50\n                ","pageSize":50}

Response
-------
Headers: {
    "date": "Mon, 13 May 2019 01:45:33 GMT",
    "alt-svc": "quic=\":443\"; ma=2592000; v=\"46,44,43,39\""
}

Fault
-------
Status code: 13
Details: Internal error encountered.
```

**Comment by BeautifulCodeStar** [comment]
That ads account was already authorized by that MCC account and it has been working for several months.

After upgrading to version1, it's always raising that exception.

No changes for code base. Ads version is only upgraded v0 to v1.

**Comment by fiboknacky** [comment]
Yes, and that what I've seen and tried to confirm with you in the previous replies. :)

From this:

```
Host: googleads.googleapis.com
Headers: {
    "x-goog-api-client": "gl-php\/7.2.17-0ubuntu0.18.04.1 gapic\/ gax\/0.38.0 grpc\/1.4.0",
    "x-goog-request-params": "customer_id=7509260745",
    "developer-token": "REDACTED",
    "login-customer-id": "992-748-0321"
}
Request: {"customerId":"7509260745","query":"SELECT campaign.id, campaign.name, metrics.ctr, metrics.clicks, metrics.impressions, metrics.average_cpc, metrics.cost_micros, metrics.average_cost, metrics.conversions\n                FROM campaign\n                LIMIT 50\n                ","pageSize":50}
```

You use customerId = 7509260745, which doesn't have any **parent account**.
So, how come you could use "992-748-0321" as a login-customer-id?

**Comment by BeautifulCodeStar** [comment]
`Parent account` meaning? That means MCC account? :)

Login customer ID is manager account ID. And this was generated by Google Ads API.

Is this wrong?

**Comment by BeautifulCodeStar** [comment]
Powerful support, please :)

**Comment by BeautifulCodeStar** [comment]
This is how it was weird? Do you think so?

**Comment by fiboknacky** [comment]
Yes, parent account is the MCC account.
It's not the wrong ID, but they're not linked together.
I've checked this three times and cannot find 7509260745 under 992-748-0321 both in a direct and indirect relationship.

It seems 7509260745 is a single account, not linked to any account.

Perhaps, there is someone removing the link that you have used before?

**Comment by BeautifulCodeStar** [comment]
I see. Good to know about it.
I'm not sure exactly about that someone as it's MCC account is not mine and is related with someone else.

In that case
Current now, I'm using that MCC account(992-748-0321) and it's developer token. Is this valid one? Is it possible you to check it for me? :)

Yes, if it's not linked together, before I've tried to reconnect to MCC and tried to generate new refresh token. Yes, generated new refresh token certainly. And It did not worked as well with that new refresh token and raised same issue.

So if it's not linked and someone was removed links, And I tried to reconnect and it's not worked.

In this case, what can I do to solve a issue? Could you please share a good idea for me?

Thank you.

**Comment by fiboknacky** [comment]
> Current now, I'm using that MCC account(992-748-0321) and it's developer token. Is this valid one? Is it possible you to check it for me? :)

As long as you've not revoked your developer token and don't attack our servers, it'll be valid. :)

> Yes, if it's not linked together, before I've tried to reconnect to MCC and tried to generate new refresh token. Yes, generated new refresh token certainly. And It did not worked as well with that new refresh token and raised same issue. So if it's not linked and someone was removed links, And I tried to reconnect and it's not worked.

I'm not sure I got what you mean, but if you have access to that account, you should be able to generate refresh tokens for that account.
If that account is not linked anymore, legally and technically, you cannot access to that account.
And that's nothing you can do for such a case.

Just let us know in [the forum](https://groups.google.com/forum/?nomobile=true#!forum/adwords-api) if you have a problem generating a new refresh token.

**Comment by BeautifulCodeStar** [comment]
I've reconnected Adwords account 7509260745 to MCC account 992-748-0321 and got new refresh token as well.

Now, is my Ads account connected to that MCC?

Even though it's connected, it's still causing that Exception `INTERNAL ERROR ENCOUNTERED`.

**Comment by BeautifulCodeStar** [comment]
Could you please check that was connected successfully for me?

**Comment by fiboknacky** [comment]
> I've reconnected Adwords account 7509260745 to MCC account 992-748-0321 and got new refresh token as well.

How did you do that? Unfortunately, I don't see they have parent-child link yet.

**Comment by BeautifulCodeStar** [comment]
I see. It's really strange.

I've generated new refresh token via terminal by using Google Ads's samples `examples\Authentication\AuthenticationInWebApplication.php`.

**Comment by BeautifulCodeStar** [comment]
![image](https://user-images.githubusercontent.com/34084225/57919347-51dcbf80-78a1-11e9-86e0-7f49bb6b45e5.png)

As you may see, My google account is saying it was allowed to access by that MCC.

**Comment by fiboknacky** [comment]
Please review [this section](https://developers.google.com/adwords/api/docs/guides/accounts-overview#managedcustomerservice).
This is what you needed to link them together as a parent-child relationship.
Have you followed those steps?

**Comment by BeautifulCodeStar** [comment]
I know those steps and did not followed.

Through OAuth2 authentication steps, it's enough to access user's Ads data, no?

**Comment by BeautifulCodeStar** [comment]
I'm confusing. Does it needs to be connected with such way? not OAuth2?

Could you please explain in details?

**Comment by fiboknacky** [comment]
It seems you don't understand well about context of [login-customer-id](https://developers.google.com/google-ads/api/docs/concepts/call-structure#login-customer-id).
Could you please review that first?

In short, if you wish to use the same settings (using MCC account as login-customer-id), you need to link them together as a parent-child relationship.

If you just need to send a request to that child account directly, you don't need [this step](https://developers.google.com/google-ads/api/docs/account-management/linking-manager-accounts).
But, you **cannot** use login-customer-id like you're trying to do right now.
You need to generate a refresh token (or access token) when logged in as a Google account that has access to Ads customer ID 7509260745.
Then, **do not pass** login-customer-id when you send a request to the API.

As the conversation is diverged from the client libraries, could you please post on [the forum](https://groups.google.com/forum/?nomobile=true#!forum/adwords-api) instead?
There are more people who could answer your questions across many time zones. :)
Note that I've just told my colleagues that the post you've created is handled here, but because I thought it was related to the client library.
Now it's more like a concept question, so it's suitable to continue the conversation there.

**Comment by BeautifulCodeStar** [comment]
`You need to generate a refresh token (or access token) when logged in as a Google account that has access to Ads customer ID 7509260745.`
Yes. I think I've mentioned few times above about it.

`As the conversation is diverged from the client libraries, could you please post on the forum instead?`
I have already posted this to the forum and one of the members is said it's related to the client library.
Yes, it's certainly client library issue as you can see through API call.

`you cannot use login-customer-id like you're trying to do right now.`
Yes, In the successful response for Google Ads API call, there is no such header `login-customer-id`.
But currently, it's generating such header with API call? This is not a issue related to the client library?

**Comment by BeautifulCodeStar** [comment]
What I'm saying is that

This google-ads-php client library is generating such unexpected login-customer-id for /google.ads.googleads.v1.services.GoogleAdsService/Search method, Hence API server is returning such exception `INTERNAL ERROR ENCOUNTERED`

Have you got what I mean?

Please let me know if anything wrong.

**Comment by BeautifulCodeStar** [comment]
I don't want to take much time.
I want to know whether this issue was related to the Google Ads client library..
If yes, I can't wait until it's fixed or I could use Google Adwords API back.

Please advise.

**Comment by BeautifulCodeStar** [comment]
This is a reply from Google Ads forum.

```
Hi,

There is currently no workaround that I am aware of for this issue. Also, the Internal error issue you encountered is more related to the client library. I would recommend that you follow up on the thread as the client library owners are better equipped to assist you for this matter.

Regards,
Dave
Google Ads API Team
```

**Comment by fiboknacky** [comment]
> you cannot use login-customer-id like you're trying to do right now. Yes, In the successful response for Google Ads API call, there is no such header login-customer-id. But currently, it's generating such header with API call? This is not a issue related to the client library?

I don't get what you mean here. Do you mean the library still generates such a header even when you don't specify anything?

Could you attach the latest response of your code?
Is it still the same as [this comment](https://github.com/googleads/google-ads-php/issues/80#issuecomment-492994885) of yours?

**Comment by BeautifulCodeStar** [comment]
Yes. exactly

I'm not trying to any direct call for Google Ads API, just using Google Ads client library as suggested.

That's a latest API log.

**Comment by BeautifulCodeStar** [comment]
And as I mentioned above, it worked well for several months but no after upgrading, May

**Comment by BeautifulCodeStar** [comment]
We have all ur clients enquiring about the fix almost everyday. Please advise.

**Comment by AnashOommen** [comment]
Just curious—have you actually provided the login-customer-id in a hyphenated manner, or is it the way the id gets logged in the file?

Could you make sure it isn’t hyphenated?

**Comment by BeautifulCodeStar** [comment]
It's provided in a hyphenated manner as suggested in google_ads.ini configuration file..
Like "XXX-XXX-XXXX"

**Comment by fiboknacky** [comment]
OK. So, I think we found two issues here:

1) login customer ID shouldn't have a hyphen. I checked the most recent google_ads_api.ini file and I don't find that example anymore. Could you please check?

2) If you still got the same API log, then you're doing things wrongly here.
As I said above, 7509260745 is not linked to any MCC accounts now. So, you CANNOT use login-customer-id.

So, in short, could you please make sure you DON'T provide any login-customer-id and try again?

**Comment by BeautifulCodeStar** [comment]
`7509260745 is not linked to any MCC accounts now`
If I understood correctly, I am not sure why you are still talking about this.

I think we are talking about account access by TOKEN, like by OAuth2 authentication.

Yes. Same issue is happening with or without dash.
And I am not trying to do any direct call and using this Google Ads PHP client library hence I can't not control any headers in API call as it's generated by library.

**Comment by BeautifulCodeStar** [comment]
Whole Ads features are not working with this issue for nearly 2 weeks. Please advice.

**Comment by AnashOommen** [comment]
Thanet, can we query the logs by developer token? That should tell us whether the internal error is from our server, or the grpc layer.

**Comment by fiboknacky** [comment]
@BeautifulCodeStar 
> If I understood correctly, I am not sure why you are still talking about this.
> I think we are talking about account access by TOKEN, like by OAuth2 authentication.

I'm afraid that you don't understand what I explain and ask you to try. 
Could you please **stop using login-customer-id** for now and try?
Again, the login-customer-id header is for the case when you want to specify the account ID you want to log in as (similar to when you log into the UI). That's valid only when you manage accounts using MCC accounts in a hierarchy.

In your case, 7509260745 IS NOT LINKED to any MCC accounts.
So, it's not related to 992-748-0321 AT ALL. (Probably they would have been linked together, but they're not now)
That's why when you still send 992-748-0321 as the login-customer-id (if I understand correctly that you've not changed anything except for a refresh token since [this reply](https://github.com/googleads/google-ads-php/issues/80#issuecomment-492994885)), it keeps failing.
It's like you're trying to set a random account as the login-customer-id header. That's not supposed to work, right?

I know you said that it worked before, but there were many changes including this login-customer-id logic, during v0 to v1 transition.
And most importantly, 7509260745 is not linked to 992-748-0321 anymore.
**What we're trying to diagnose now is NOT RELATED TO refresh and access tokens, but rather the login-customer-id header. So please try deleting the [login-customer-id line in the config file](https://github.com/googleads/google-ads-php/blob/master/examples/Authentication/google_ads_php.ini#L9) first and see what happens, so I can help you in the next step. :)**

@AnashOommen 
I did last week and the error is authentication error because of the permission denied.
That's because, as I'm trying to say several times above, login-customer-id cannot be used in this case because 7509260745 is just a single account not managed by any MCC accounts.
If the login-customer-id header is to be specified, it should be at least 7509260745 itself.
It's rather not specified at all in this case.

**Comment by BeautifulCodeStar** [comment]
Glad it's solved! Thank you! @fiboknacky

**Comment by BeautifulCodeStar** [comment]
Now I can close this issue. Thanks for your time, all! 👍

**Comment by chiragvels** [comment]
I have started facing issue with this as I tried to get customer clients. Till last monday it was working fine but from yesterday I was facing same.

Also when I use pagination it is giving me same data for every page.

ApiException was thrown with message '{ "message": "Internal error encountered.", "code": 13, "status": "INTERNAL", "details": [ { "@type": "request-id", "data": "E_HdP_3LHJKibFGRtI5h-Q" } ] }'. 

I have log with me i can share privately If required.

Thanks,
